### PR TITLE
Add training scripts (#22)

### LINKED ISSUE #22: Weights from MAISI in need of fine-tuning
It was decided that foundation models from [here](https://huggingface.co/nvidia/NV-Generate-CT) were going to be utilized. Since the data normalization strategy is being changed the VAE has to be retrained, the U-Net needs fine-tuning as the original network definition does not accept direct conditioning and is instead received from ControlNet in the form of residuals. The model also needs to adapt to the new VAE. This should be done:
- Train the VAE.
- After the VAE is set in stone, train the weights of U-Net with a warm-start from the foundation model weights.
- Document the training scripts along with necessary information.

### PR DESCRIPTION
**Pull request by Couch-bit**
## Description

This PR adds the training scripts which were used for training both the VAE-GAN and the U-Net. It also adds a new doc describing critical information relating to these scripts. Keep in mind that there is a placeholder where the epoch index containting the final Rectified Flow weights would be. Picking that will be a part of a separate issue.

## Related Issue

Closes #22

## Proposed Changes

- Trains the VAE.
- Trains the weights of U-Net with a warm-start from the foundation model weights.
- Documents the training scripts along with necessary information.

### CODE REVIEW

File: src/training/helpers/config.py [hunk 1]
```diff
@@ -0,0 +1,487 @@
...
+
+
+@dataclass
+class MaisiTrainingConfig:
+    """
+    Store configuration and architectural parameters for the MAISI CT training pipeline.
+
+    This class maintains, normalizes, and validates spatial, intensity, hyperparameter,
+    and filesystem structural requirements for both VAE and U-Net optimization tasks.
+
```
Comment (MagicWiqqu) [inline src/training/helpers/config.py:16]: Maybe add `Assumptions` and `Pipeline stages` for consistency between this config, and the pipeline one? Training also has stages of running, etc.

File: src/training/helpers/data_loading.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+class LoadProcessedTensord(MapTransform):
+    """
+    Dictionary-based transform to load preprocessed PyTorch tensor files (.pt).
+
+    Parameters
+    ----------
+    keys : KeysCollection
+        Keys of the corresponding items to be transformed.
+    allow_missing_keys : bool
+        Don't raise exception if key is missing.
```
Comment (MagicWiqqu) [inline src/training/helpers/data_loading.py]: Documentation lacks the `Raises` part, despite having an exception.

File: src/training/helpers/data_loading.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+                # weights_only=True helps with security/speed
+                d[key] = torch.load(d[key], weights_only=True)
+            except Exception as err:
+                raise RuntimeError(f"Failed to load preprocessed tensor file from path: {d[key]}") from err
+
+        return d
+
+
+def get_ct_preprocessing_transform(config: MaisiTrainingConfig) -> Compose:
+    """
```
Comment (MagicWiqqu) [inline src/training/helpers/data_loading.py:24]: The same function exists within /pipeline/data/prepare_test_data.py. It is redundant, no?

File: src/training/helpers/datasets.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    def __init__(self, data_dir: str | Path) -> None:
+        self.data_dir = Path(data_dir)
+
+        if not self.data_dir.exists():
+            raise FileNotFoundError(f"The provided data directory does not exist: {self.data_dir}")
+
+        self.pairs = []
+
+        # Regex to capture patient_id and fraction_id
+        # Handles both numeric and alphanumeric patient IDs
```
Comment (MagicWiqqu) [inline src/training/helpers/datasets.py]: What do you mean by alphanumeric? Isn't this only numeric?

File: src/training/helpers/model_init.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+    try:
+        checkpoint = torch.load(config.rflow_foundation_weights, weights_only=False)
+    except Exception as err:
+        raise RuntimeError(
+            f"Failed to load foundation weights checkpoint from {config.rflow_foundation_weights}"
+        ) from err
+
+    if "unet_state_dict" not in checkpoint:
+        raise KeyError(
+            f"The loaded checkpoint at {config.rflow_foundation_weights} is missing the required 'unet_state_dict' key."
```
Comment (MagicWiqqu) [inline src/training/helpers/model_init.py]: Delete dot from the exception.

File: src/training/helpers/model_init.py [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+        )
+
+    rflow_state_dict = checkpoint["unet_state_dict"]
+    first_layer_key = "conv_in.conv.weight"
+
+    if first_layer_key not in rflow_state_dict:
+        raise KeyError(f"The key '{first_layer_key}' was not found within the loaded 'unet_state_dict'.")
```
Comment (MagicWiqqu) [inline src/training/helpers/model_init.py]: Delete dot from the exception.

File: src/training/helpers/model_init.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+
+    model_state = model.state_dict()
+    if first_layer_key not in model_state:
+        raise KeyError(f"The key '{first_layer_key}' is missing from the initialized UNet model architecture.")
```
Comment (MagicWiqqu) [inline src/training/helpers/model_init.py]: Gemini detected, dot rejected.

File: src/training/prepare_data.py [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+
+    for path in tqdm(ct_paths["0"]["val"], desc="Processing validation data"):
+        process_ct(Path(path), transform, config, True)
+
+
+def main() -> None:
+    """
+    The main entry point for processing CT scans from NIfTI format to VAE training ready .pt files.
+
+    The only dependency this has is the presence of the full dataset in its directory
```
Comment (MagicWiqqu) [inline src/training/prepare_data.py:126]: I do not think so. Creating MaisiTrainingConfig validates the weights available and such.

File: src/training/helpers/data_loading.py [hunk 9]
```diff
@@ -0,0 +1,241 @@
...
+    Compose
+        A MONAI Compose object containing the preprocessing sequence.
+    """
+
+    return Compose(
+        [
+            LoadImaged(keys=["image"]),
+            EnsureChannelFirstd(keys=["image"]),
+            Orientationd(keys=["image"], axcodes="RAS", labels=None),
+            SpatialPadd(
```
Comment (MagicWiqqu) [inline src/training/helpers/data_loading.py:46]: We should first Scale, and then Pad. If you pad and introduce zeroes, after Scaling they become like mid-intensity tissue, and not air.

File: src/training/train_rflow.py [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+        start_epoch = 0
+
+    # Make sure the model works correctly with DDP
+    if is_distributed:
+        model = DDP(model, device_ids=[local_rank], output_device=local_rank)
+
+    model.compile()
+
+    # The spacing and class tensors are consistent so they can be defined here
+    class_labels = torch.tensor([2], dtype=torch.long).to(device)
```
Comment (MagicWiqqu) [inline src/training/train_rflow.py:251]: In inference, we have [0], and not [2]. It should match, no?

Comment (MagicWiqqu) [reply src/training/train_rflow.py:251]: Also in the inference condition is not scaled if I remember correctly (here line 292)

Comment (MagicWiqqu) [reply src/training/train_rflow.py:251]: Also shouldn't this be in MaisiTrainingConfig to be fair?

File: src/training/train_vae.py [hunk 11]
```diff
@@ -0,0 +1,527 @@
...
+from src.training.helpers.ema import EMA
+from src.training.helpers.losses import adv_loss, intensity_loss, kl_loss, perceptual_loss
+from src.training.helpers.model_init import discriminator_init, vae_init
+
+# Optional optimizations
+torch.set_float32_matmul_precision("high")
+torch.backends.cudnn.benchmark = True
+
+
+def save_checkpoint(state: dict[str, Any], file: str | Path) -> None:
```
Comment (MagicWiqqu) [inline src/training/train_vae.py:31]: Isn't this redundant and also in train_rflow.py?

File: src/training/train_vae.py [hunk 12]
```diff
@@ -0,0 +1,527 @@
...
+            optimizer_d.zero_grad(set_to_none=True)
+
+            # Normalize the KL by the number of elements in the batch
+            norm_factor = images.numel()
+
+            # Get loss values
+            reconstruction, z_mu, z_sigma = cast(tuple[torch.Tensor, torch.Tensor, torch.Tensor], autoencoder(images))
+
+            # Train Generator
+            logits_fake = discriminator(reconstruction.contiguous())[-1]
```
Comment (MagicWiqqu) [inline src/training/train_vae.py:361]: Shouldn't we freeze the parameters till line 371?

File: src/training/train_vae.py [hunk 13]
```diff
@@ -0,0 +1,527 @@
...
+                    # Keep the best model weights on CPU to avoid uncessary VRAM usage
+                    raw_model = cast(AutoencoderKlMaisi, autoencoder.module) if is_distributed else autoencoder
+                    best_model_wts = {k: v.cpu() for k, v in raw_model.state_dict().items()}
+            else:
+                epochs_no_improve += config.val_interval
+                # EVERY rank breaks simultaneously
+                if epochs_no_improve >= config.patience:
+                    break
+
+            ema.restore(autoencoder)
```
Comment (MagicWiqqu) [inline src/training/train_vae.py:480]: If early stopping is triggered, this will not happen, no?

File: src/training/train_vae.py [hunk 14]
```diff
@@ -0,0 +1,527 @@
...
+            # EARLY STOPPING LOGIC (Runs on all ranks!)
+            if val_epoch_losses_float["loss"] < best_val_loss:
+                best_val_loss = val_epoch_losses_float["loss"]
+                epochs_no_improve = 0
+
+                # Only Rank 0 saves the actual weights to memory
+                if rank == 0:
+                    # Keep the best model weights on CPU to avoid uncessary VRAM usage
+                    raw_model = cast(AutoencoderKlMaisi, autoencoder.module) if is_distributed else autoencoder
+                    best_model_wts = {k: v.cpu() for k, v in raw_model.state_dict().items()}
```
Comment (MagicWiqqu) [inline src/training/train_vae.py]: best_model_wts = {
    k: v.detach().cpu().clone()
    for k, v in raw_model.state_dict().items()
}

Is safer in case someone is training on CPU (crazy)

**Review by MagicWiqqu — CHANGES_REQUESTED** [review]
Please look over my comments. Reach out if anything is unclear - it was a bit hastily done due to the fact that I wanted to finish before Friday for you.

File: src/training/helpers/config.py (hunk 1 above)
Comment (Couch-bit) [reply src/training/helpers/config.py:16]: Okay, will do

File: src/training/helpers/data_loading.py (hunk 2 above)
Comment (Couch-bit) [reply src/training/helpers/data_loading.py]: Noted

File: src/training/helpers/data_loading.py (hunk 3 above)
Comment (Couch-bit) [reply src/training/helpers/data_loading.py:24]: The config is defined slightly differently so I would like to keep them separate.

File: src/training/helpers/datasets.py (hunk 4 above)
Comment (Couch-bit) [reply src/training/helpers/datasets.py]: Should be numeric, I simplified as the regex originally also worked for alphanumeric characters.

File: src/training/helpers/model_init.py (hunk 5 above)
Comment (Couch-bit) [reply src/training/helpers/model_init.py]: Yes

File: src/training/helpers/model_init.py (hunk 6 above)
Comment (Couch-bit) [reply src/training/helpers/model_init.py]: Yes

File: src/training/helpers/model_init.py (hunk 7 above)
Comment (Couch-bit) [reply src/training/helpers/model_init.py]: Yes

File: src/training/prepare_data.py (hunk 8 above)
Comment (Couch-bit) [reply src/training/prepare_data.py:126]: Technically yes, but it does not check if `src/data_full` is fully populated as only the user knows how much data exactly should be there.

File: src/training/train_rflow.py (hunk 10 above)
Comment (Couch-bit) [reply src/training/train_rflow.py:251]: I don't think this should be in the config, as 2 is simply a CT without contrast. 0 is Unknown so 2 is better.

File: src/training/helpers/data_loading.py (hunk 9 above)
Comment (Couch-bit) [reply src/training/helpers/data_loading.py:46]: Technically you are right, but I don't think it really matters that much what we pad the edges with. Keep in mind that the edges may not be air, the missing depth may actually be largely the stomach.

File: src/training/train_vae.py (hunk 14 above)
Comment (Couch-bit) [reply src/training/train_vae.py]: I can do that, I haven't run this on CPU and that's kind of unpractical anyway, but better for the code to be more robust I suppose.

File: src/training/train_vae.py (hunk 13 above)
Comment (Couch-bit) [reply src/training/train_vae.py:480]: Yes, but it does not need to happen.

File: src/training/train_vae.py (hunk 12 above)
Comment (Couch-bit) [reply src/training/train_vae.py:361]: Technically, we should, but MAISI had the gradients leak, so I just did it the same. I didn't notice it at first. I don't want to change this and rerun, there is also a possibility that those gradients actually act as a regularizer and help for all we know.

File: src/training/train_vae.py (hunk 11 above)
Comment (Couch-bit) [reply src/training/train_vae.py:31]: Technically it's the same function, but operates on different assumptions (models) and is paired with a different `load_checkpoint`. That's my rationale for not moving this to a separate file.

File: src/training/train_vae.py (hunk 13 above)
Comment (Couch-bit) [reply src/training/train_vae.py:480]: It's only needed for the next epoch, if there is no next epoch it doesn't matter.

File: src/training/train_vae.py (hunk 12 above)
Comment (Couch-bit) [reply src/training/train_vae.py:361]: I will comment on it in the code.

File: src/training/helpers/data_loading.py (hunk 9 above)
Comment (Couch-bit) [reply src/training/helpers/data_loading.py:46]: I will comment on this in the code to avoid future confusion.

File: src/training/train_vae.py (hunk 14 above)
Comment (Couch-bit) [reply src/training/train_vae.py]: Also there is a typo in the comment here, so I will fix that too.

File: src/training/helpers/data_loading.py (hunk 2 above)
Comment (Couch-bit) [reply src/training/helpers/data_loading.py]: This class is also used in the pipeline, so I should add it to `src/data_utils` instead and have it imported where needed.

File: src/training/helpers/data_loading.py (hunk 3 above)
Comment (MagicWiqqu) [reply src/training/helpers/data_loading.py:24]: As you wish - I just noted it in case it was not on purpose.

File: src/training/train_vae.py (hunk 11 above)
Comment (MagicWiqqu) [reply src/training/train_vae.py:31]: Okay, then it can stay here.

File: src/training/train_rflow.py (hunk 10 above)
Comment (MagicWiqqu) [reply src/training/train_rflow.py:251]: Okay, then it must be changed in the inference. May you add it to the issue (#37)? You may also add anything else we need to change.

File: src/training/train_vae.py (hunk 13 above)
Comment (MagicWiqqu) [reply src/training/train_vae.py:480]: Ok.

File: src/training/prepare_data.py (hunk 8 above)
Comment (MagicWiqqu) [reply src/training/prepare_data.py:126]: Ok, it can be phrased like that, then.

**Review by MagicWiqqu — APPROVED** [review]
LGTM
